File: internal/web/memory_evidence.go
package web

import (
	"context"
	"encoding/json"
	"net/http"

	"github.com/davidadel66/evie/internal/memory"
)

// MemoryEvidenceInspector resolves the exact references retained with a
// request, applying the caller's current access before returning source text.
type MemoryEvidenceInspector interface {
	InspectMemoryEvidence(context.Context, memory.ScopeContext, []memory.RetrievalReference) ([]memory.RetrievalInspection, error)
}

func (s *Server) handleMemoryEvidence(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Cache-Control", "no-store")
	var request struct {
		SessionID  memory.SessionID `json:"sessionId"`
		SnapshotID memory.EventID   `json:"snapshotId"`
		AnswerID   memory.EventID   `json:"answerId"`
	}
	if status, err := decodeManagementJSON(w, r, &request); err != nil || request.SessionID == "" || (request.SnapshotID == "" && request.AnswerID == "") {
		if status == 0 {
			status = http.StatusBadRequest
		}
		managementJSONError(w, status, "invalid_memory_evidence", "A selected conversation and request reference are required.")
		return
	}
	s.sessionMu.RLock()
	defer s.sessionMu.RUnlock()
	if s.selectingSession || s.activeSession.ID != request.SessionID || s.session == nil {
		managementJSONError(w, http.StatusConflict, "memory_session_changed", "The selected conversation changed. Open its sources again.")
		return
	}
	var events []memory.Event
	var err error
	if history, ok := s.memoryEvidence.(MemoryEvidenceHistory); ok {
		events, err = history.LoadEvents(r.Context(), request.SessionID)
	} else {
		events, err = s.session.HistoryEvents(r.Context())
	}
	if err != nil {
		managementJSONError(w, http.StatusConflict, "memory_evidence_unavailable", "The original request could not be inspected.")
		return
	}
	records, err := memoryRequestRecords(events)
	if err != nil {
		managementJSONError(w, http.StatusConflict, "memory_evidence_unavailable", "The original request could not be inspected.")
		return
	}
	result, err := inspectMemoryRequests(r.Context(), s.memoryEvidence, s.activeSession.ScopeContext(), records, request.SnapshotID, request.AnswerID)
	if err != nil {
		managementJSONError(w, http.StatusUnprocessableEntity, "memory_evidence_unavailable", "The original evidence is unavailable under current access.")
		return
	}
	if result != nil {
		writeJSON(w, http.StatusOK, result)
		return
	}
	managementJSONError(w, http.StatusNotFound, "memory_evidence_unavailable", "No original evidence reference is available for this request.")
}

type memoryActivity struct {
	SnapshotID      memory.EventID `json:"snapshotId"`
	RequestStatus   string         `json:"requestStatus,omitempty"`
	Iteration       int            `json:"iteration,omitempty"`
	AnswerID        memory.EventID `json:"answerId,omitempty"`
	Status          string         `json:"status"`
	AcceptedCount   int            `json:"acceptedCount"`
	ExcerptCount    int            `json:"excerptCount"`
	HistoricalCount int            `json:"historicalCount,omitempty"`
	RetiredCount    int            `json:"retiredCount,omitempty"`
	ConflictCount   int            `json:"conflictCount,omitempty"`
}

func projectMemoryActivity(event memory.Event) (*memoryActivity, error) {
	var snapshot memory.ContextSnapshotPayload
	if len(event.Payload) == 0 {
		return nil, nil
	}
	if err := json.Unmarshal(event.Payload, &snapshot); err != nil {
		return nil, err
	}
	if snapshot.Memory == nil {
		return nil, nil
	}
	activity := &memoryActivity{SnapshotID: event.ID, RequestStatus: "prepared", Iteration: snapshot.Iteration, Status: snapshot.Memory.Status}
	for _, ref := range snapshot.Memory.Evidence {
		switch ref.Kind {
		case memory.RetrievalAcceptedMemory:
			activity.AcceptedCount++
		case memory.RetrievalConversationExcerpt:
			activity.ExcerptCount++
		}
		if ref.Intent == memory.RetrievalHistorical {
			activity.HistoricalCount++
		}
		if ref.Status == memory.SemanticStatusRetired || ref.CurrentStatus == memory.SemanticStatusRetired {
			activity.RetiredCount++
		}
		if len(ref.Conflicts) > 0 {
			activity.ConflictCount++
		}
	}
	return activity, nil
}

func (e *sseEvents) MemoryRetrieved(event memory.Event) {
	activity, err := projectMemoryActivity(event)
	if err != nil {
		e.Error("Memory activity could not be displayed. Reload the conversation.")
		return
	}
	if activity != nil {
		e.emit("memory_activity", activity)
	}
}

File: internal/web/ui/src/api/memoryEvidence.ts
import type { SemanticClaim, SemanticObjectInspection } from "./memory";

export type MemoryActivityData = {
  snapshotId: string;
  status: string;
  acceptedCount: number;
  excerptCount: number;
  historicalCount?: number;
  retiredCount?: number;
  conflictCount?: number;
};

export type MemorySourceReference = {
  source_link_id?: string;
  event_id: string;
  event_part: string;
  session_id: string;
  scope_key: string;
  authority: string;
  observed_at: string;
  locator_kind: string;
  locator_value: string;
  evidence_sha256: string;
};

export type MemoryReference = {
  id: string;
  kind: string;
  claim_id?: string;
  claim_operation_id?: string;
  as_known_at: string;
  valid_at: string;
  scope_key: string;
  status: string;
  current_status?: string;
  intent?: string;
  valid_at_constrained?: boolean;
  correction_mode?: string;
  current_correction_mode?: string;
  conflicts?: SemanticObjectInspection["conflicts"];
  related_claim_ids?: string[];
  paths: string[];
  sources: MemorySourceReference[];
};

export type MemoryEvidenceReceipt = {
  sessionId: string;
  snapshotId: string;
  version: string;
  status: string;
  evidence: {
    reference: MemoryReference;
    available: boolean;
    current_status: string;
    evidence?: {
      text: string;
      claim?: Pick<SemanticClaim, "transaction_time" | "valid_time">;
      effective_valid_time?: SemanticClaim["valid_time"];
      current_status?: string;
      correction_mode?: string;
      current_correction_mode?: string;
      conflicts?: SemanticObjectInspection["conflicts"];
      related_claim_ids?: string[];
      sources: {
        source_link_id?: string;
        event_id: string;
        session_id: string;
        source_scope_key: string;
        authority: string;
        actor?: string;
        observed_at: string;
        evidence: string;
        locator_kind: string;
        locator_value: string;
        evidence_sha256: string;
      }[];
    };
  }[];
};

export async function inspectMemoryEvidence(sessionId: string, snapshotId: string, signal?: AbortSignal): Promise<MemoryEvidenceReceipt> {
  const response = await fetch("/api/memory/evidence", {
    method: "POST", headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ sessionId, snapshotId }), cache: "no-store", signal,
  });
  const value = await response.json() as MemoryEvidenceReceipt & { error?: string };
  if (!response.ok) throw new Error(value.error ?? "Original memory evidence is unavailable.");
  if (value.sessionId !== sessionId || value.snapshotId !== snapshotId) throw new Error("The selected conversation changed. Open its sources again.");
  return { ...value, evidence: value.evidence ?? [] };
}
